adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java

```diff
@@ -354,6 +369,44 @@ protected final void doStart(final Promise<Void> startPromise) {
             .onComplete(startPromise);
     }
 
+    private void registerDeviceAndTenantChangeNotificationConsumers() {
+        NotificationEventBusSupport.registerConsumer(vertx, DeviceChangeNotification.TYPE,
+                notification -> {
+                    if (LifecycleChange.DELETE.equals(notification.getChange())
+                            || (LifecycleChange.UPDATE.equals(notification.getChange()) && !notification.isEnabled())) {
+                        final String reason = LifecycleChange.DELETE.equals(notification.getChange()) ? "device deleted"
+                                : "device disabled";
+                        closeDeviceConnectionsOnDeviceOrTenantChange(
+                                connectedDevice -> connectedDevice.getTenantId().equals(notification.getTenantId())
+                                        && connectedDevice.getDeviceId().equals(notification.getDeviceId()),
+                                reason);
+                    }
+                });
+        NotificationEventBusSupport.registerConsumer(vertx, AllDevicesOfTenantDeletedNotification.TYPE,
+                notification -> closeDeviceConnectionsOnDeviceOrTenantChange(
+                        connectedDevice -> connectedDevice.getTenantId().equals(notification.getTenantId()),
+                        "all devices of tenant deleted"));
+        NotificationEventBusSupport.registerConsumer(vertx, TenantChangeNotification.TYPE,
+                notification -> {
+                    if (LifecycleChange.DELETE.equals(notification.getChange())
+                            || (LifecycleChange.UPDATE.equals(notification.getChange()) && !notification.isEnabled())) {
+                        final String reason = LifecycleChange.DELETE.equals(notification.getChange()) ? "tenant deleted"
+                                : "tenant disabled";
+                        closeDeviceConnectionsOnDeviceOrTenantChange(
+                                connectedDevice -> connectedDevice.getTenantId().equals(notification.getTenantId()),
+                                reason);
+                    }
+                });
+    }
+
+    private void closeDeviceConnectionsOnDeviceOrTenantChange(final Predicate<Device> deviceMatchPredicate, final String reason) {
+        final List<MqttDeviceEndpoint> deviceEndpoints =  connectedAuthenticatedDeviceEndpoints.stream()
+                .filter(ep -> deviceMatchPredicate.test(ep.getAuthenticatedDevice()))
+                .collect(Collectors.toList());
+        // sendDisconnectedEvent param false here - sending an event for a deleted/disabled tenant/device would fail
+        deviceEndpoints.forEach(ep -> ep.close(reason, false));
```

Comment:
How about doing this iteration over thousands of entries on a worker thread? Even more so as we are iterating twice ...

Reply:
That would make things more complicated here, possibly requiring extra synchronization.
I would rather deal with the MQTT client in the event loop thread here.
How about letting the processing loop be decoupled by invoking each `ep.close()` invocation inside `context.runOnContext()`?